taskcat/cli_core.py

```diff
@@ -21,83 +23,205 @@ def __init__(self, expression):
-                # and invoke method
-                LOG.debug("Creating parser for function {}".format(name))
-                parser = self.create_argparse(method)
-                kwargs = vars(parser.parse_args(arguments))
-                LOG.debug(kwargs)
-                method(**kwargs)
-                return
-        # Invalid action provided, raise InvalidActionError
-        raise InvalidActionError(
-            "Invalid sub-subcommand {} for module {}".format(
-                action, self.module_name.lower()
-            )
-        )
-
-    def get_methods(self):
-        """Return list of available methods"""
-        LOG.debug("Returns list of methods for module " + self.module_name)
-        return inspect.getmembers(self.module, predicate=inspect.ismethod)
-
-    def create_argparse(self, method):
-        LOG.debug("Method doc string -> " + str(method.__doc__))
-        LOG.debug("whose signature is -> {}".format(inspect.signature(method)))
-        # Create arg parser for module.method with parameters as options
-        # By default, all method parameters are required
-        arg_parser = argparse.ArgumentParser(
-            description=str(method.__doc__),
-            usage=f"%(prog)s {self.module_name.lower()} {method.__name__} [options]",
-        )
-        optional_group = arg_parser._action_groups.pop()
-        required_group = arg_parser.add_argument_group("required arguments")
-
-        arg_type = type("str")
-        required_switch = True
-
-        sig = inspect.signature(method)
-        for param in sig.parameters.values():
-            LOG.debug(
-                "{} is of type {} with default value {}".format(
-                    param.name, param.annotation, param.default
+    USAGE = "{prog}{global_opts}{command}{command_opts}{subcommand}{subcommand_opts}"
+
+    def __init__(self, prog_name, module_package, description, version=None, args=None):
+        self.name = prog_name
+        self.module_package = module_package
+        self._modules = self._get_plugin_modules()
+        self.args = {"global": args if args is not None else [], "commands": {}}
+        self._build_args()
+        self.command_parser = None
+        self.subcommand_parsers = {}
+        self.parser = self._build_parser(description, version)
+        self.parsed_args = []
+
+    def _build_args(self):
+        for name, module in self._modules.items():
+            params = self._get_params(module)
+            self.args["commands"][name] = {"args": params, "subcommands": {}}
+            for method_name, method_function in self._get_class_methods(module):
+                if not method_name.startswith("_"):
+                    params = self._get_params(method_function)
+                    self.args["commands"][name]["subcommands"][method_name] = params
+
+    @staticmethod
+    def _get_class_methods(module):
+        methods = inspect.getmembers(module, predicate=inspect.isfunction)
```

Comment:
Thinking we should add a check here... look for `method.cli_module` being True. Discuss?

Reply:
not sure I follow, if you mean we should have a way to tell if a method is supposed to be exposed on the cli, the thinking at the moment is that these cli modules will be wrappers whose sole purpose is to expose a cli v(as per the lint example), so not sure if this is needed.